llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/VectorExt/IR/VectorExtAttrs.td

```diff
@@ -99,5 +99,172 @@ def LayoutAttr : IREEVectorExt_Attr<"Layout",
+      - duplicate_order: Ordering of duplicates, from outermost to innermost
+
+    Finally, threads are distributed in a single duplicate. The thread
+    distribution is represented by:
+
+      - threads_per_duplicate: Sizes of this level of tiling
+      - thread_order: Ordering of dimensions, from outermost to innermost
+
+    Examples of thread distribution over a 8x4 shape:
+
+    {
+      batches_per_subgroup = [2, 1]
+      duplicates_per_batch = [2, 2]
+      threads_per_duplicate = [2, 2]
 
+      batch_order = [0, 1]
+      duplicate_order = [0, 1]
+      thread_order = [1, 0]
+    }
+
+    Distributed tile:
+
+    {
+      [0 2]|[0 2]      0,1,2,3 --> thread ids
+      [1 3]|[1 3]      
+      ------------     [x z]   --> a single duplicate tile
+      [0 2]|[0 2]      [y w]
+      [1 3]|[1 3]
+    }{
+      [0 2]|[0 2]      { ... } --> a single batch tile
+      [1 3]|[1 3]
+      ------------
+      [0 2]|[0 2]
+      [1 3]|[1 3]
+    }
+
+    So, the thread distribution looks like:
+
+    [0 2 0 2]
+    [1 3 1 3]
+    [0 2 0 2]
+    [1 3 1 3]
+    [0 2 0 2]
+    [1 3 1 3]
+    [0 2 0 2]
+    [1 3 1 3]
+
+    3. Elements per Thread
+
+    The final level of tiling, representing the minimum shape of vector that
+    is treated as an atom.
+
+    The elements are placed contigiously with their shape and ordering
+    determined by:
+      - `elements_per_thread`: Sizes of this level of tiling
+      - `element_order`: Ordering of dimensions, from outermost to innermost
+  }];
+
+  let parameters = (ins
+    ArrayRefParameter<"int64_t", "subgroups_per_workgroup">:$subgroupsPerWorkgroup,
+    ArrayRefParameter<"int64_t", "subgroup_order">:$subgroupOrder,
+
+    ArrayRefParameter<"int64_t", "batches_per_subgroup">:$batchesPerSubgroup,
+    ArrayRefParameter<"int64_t", "batch_order">:$batchOrder,
+
+    ArrayRefParameter<"int64_t", "duplicates_per_batch">:$duplicatesPerBatch,
+    ArrayRefParameter<"int64_t", "duplicate_order">:$duplicateOrder,
+
+    ArrayRefParameter<"int64_t", "threads_per_duplicate">:$threadsPerDuplicate,
+    ArrayRefParameter<"int64_t", "thread_order">:$threadOrder,
+
+    ArrayRefParameter<"int64_t", "elements_per_thread">:$elementsPerThread,
+    ArrayRefParameter<"int64_t", "element_order">:$elementOrder
+  );
+
+  // TODO: add custom parser/printer and builder to elide default value array
+  // refs.
+  let assemblyFormat = [{
+    `<` `subgroups_per_workgroup` `=` `[` $subgroupsPerWorkgroup `]` `,`
+        `batches_per_subgroup`    `=` `[` $batchesPerSubgroup `]` `,`
+        `duplicates_per_batch`    `=` `[` $duplicatesPerBatch `]` `,`
+        `threads_per_duplicate`   `=` `[` $threadsPerDuplicate `]` `,`
+        `elements_per_thread`     `=` `[` $elementsPerThread `]` `,`
+
+        `subgroup_order`          `=` `[` $subgroupOrder `]` `,`
+        `batch_order`             `=` `[` $batchOrder `]` `,`
+        `duplicate_order`         `=` `[` $duplicateOrder `]` `,`
```

Comment:
Do we need a duplicate order? Can we just let the duplicate order match the batch order? IIUC duplicate order is just to tell us which duplicate sizes match to which batch dimensions. This is in contrast to subgroup/batch/thread/element order which corresponds to an index type, unless I'm not understanding duplication here properly.

Reply:
Not really. Duplication is a completely different level of tiling than batches. For example:

```
subgroupX:

order=[1, 0]

batch0 batch1
batch2 batch3

batchX:

order[1, 0]

dup0 dup2
dup1 dup3

dupX:

order[0, 1]

thread0 thread1
thread2 thread3
```

Here, batches and duplicates are distributed in a different order.

Reply:
Ok let me step back a second and ask this. Take a broadcasted add
```mlir
%1 = vector.broadcast %0 : vector<1xf32> to vector<1024xf32>
arith.addf %1, %2 : vector<1024xf32>
```
And I want to distribute this to 16 subgroups with size 64 (lets assume one add per thread to make things easy). For `%2`, I would assume the layout I want is
```
  subgroups_per_workgroup = [16],
  batches_per_subgroup = [1],
  duplicates_per_batch = [1],
  threads_per_duplicate = [64],
  elements_per_thread = [1],

  subgroup_order = [0],
  batch_order = [0],
  duplicate_order = [0],
  thread_order = [0],
  element_order = [0]>
```
What layout do I use for `%0`, the pre-broadcast vector?

Reply:
I would use:

```
  subgroups_per_workgroup = [1],
  batches_per_subgroup = [1],
  duplicates_per_batch = [1],
  threads_per_duplicate = [1],
  elements_per_thread = [1],

  subgroup_order = [0],
  batch_order = [0],
  duplicate_order = [0],
  thread_order = [0],
  element_order = [0]>
```

This means that all subgroups and workgroups own the same element. There is an implicit rounding.
If you have 8 subgroups, but are distributing only on 4, then each element is being processed by two subgroups.

Reply:
Might be worth rethinking the naming of `subgroups_per_workgroup` and `threads_per_duplicate`, but at a minimum please add a note about this implicit broadcasting to the layout description.

Reply:
added